### PR TITLE
71 add teraxion tfn driver

### PR DESCRIPTION
**Pull request by rbudhrani**

### CODE REVIEW

**Review by pieterjbotma — DISMISSED** [review]

File: qmi/instruments/teraxion/tfn.py [hunk 1]
```diff
@@ -0,0 +1,667 @@
...
+
+        Returns:
+            the firmware version.
+        """
+        _logger.info("[%s] Getting firmware version of instrument", self._name)
+        self._check_is_open()
+        # get response
+        resp = self._read(Teraxion_TFNCommand_GetFirmwareVersion)
+        # get the data after the status bytes and ignore the null bytes
+        firmware_version = resp[self.LEN_STATUS_BYTES:]
```
[the remark this answers is not in the document]
Comment (rbudhrani) [reply qmi/instruments/teraxion/tfn.py:379]: ah that's much cleaner, will do that

[the remark this answers is not in the document]
Comment (rbudhrani) [reply qmi/instruments/teraxion/tfn.py]: I was able to do it for a few, but for some there's no fixed length-the manufacturer name, model, serial and manufacturing date. I have to read everything after 4 bytes and use the null byte as a stop marker.
For the CMD_GET_NOMINAL_SETTINGS command the manual says that 12 bytes are returned but actually 16 are, so I had to add a dummy 4 bytes at the end.

[the remark this answers is not in the document]
Comment (rbudhrani) [reply qmi/instruments/teraxion/tfn.py]: I also noticed a bug in creation of the read command `_make_read_command`. I was sending the `num_received_bytes` as decimal but they should be sent as hex. It seems that the driver did not care about the format, so everything still worked with the decimal representation

[the remark this answers is not in the document]
Comment (rbudhrani) [reply qmi/instruments/teraxion/tfn.py]: I also added some sleep time for the commands as defined by the manual. Turns out sending commands in quick succession with the process delay does cause issues

**Review by pieterjbotma — APPROVED** [review]

**Pull request merged**